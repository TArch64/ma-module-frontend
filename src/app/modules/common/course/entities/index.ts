export { Course, ICourseJSON } from './course';
export { Mentor, IMentorJSON } from './mentor';
export { Student, IStudentJSON } from './student';
