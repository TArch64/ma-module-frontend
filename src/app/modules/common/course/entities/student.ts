import { IUserJSON, User } from '@common/auth';

export interface IStudentJSON {
    id: string;
    user: IUserJSON;
}

export class Student {
    public static fromJSON(json: IStudentJSON): Student {
        return new Student(json.id, User.fromJSON(json.user));
    }

    private constructor(
        public readonly id: string,
        public readonly user: User
    ) {}
}
