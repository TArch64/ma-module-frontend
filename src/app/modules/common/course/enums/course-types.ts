export enum CourseTypes {
    GENERAL = 'GENERAL',
    ADDITIONAL = 'ADDITIONAL'
}
