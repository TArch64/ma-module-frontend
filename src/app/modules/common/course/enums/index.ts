export { CourseTypes } from './course-types';
export { MentorRoles } from './mentor-roles';
