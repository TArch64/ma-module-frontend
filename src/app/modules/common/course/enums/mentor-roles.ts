export enum MentorRoles {
    LEAD = 'LEAD',
    MENTOR = 'MENTOR'
}
