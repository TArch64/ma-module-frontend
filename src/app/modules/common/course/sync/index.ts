export {CoursesRepositorySync, ICoursesRepositorySync} from './courses-repository.sync';
