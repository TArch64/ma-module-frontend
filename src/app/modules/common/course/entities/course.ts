import {CourseTypes} from "../enums";

export interface ICourseJSON {
    id: string,
    name: string,
    type: CourseTypes
}

export class Course {
    public static fromJSON(json: ICourseJSON): Course {
        return new Course(json.id, json.name, json.type);
    }

    protected constructor(
        public readonly id: string,
        public readonly name: string,
        public readonly type: CourseTypes
    ) {}
}
