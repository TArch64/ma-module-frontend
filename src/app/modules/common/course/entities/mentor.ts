import {IUserJSON, User} from "@common/auth";
import {MentorRoles} from "../enums";
import {IPrototype} from "@common/core";

export interface IMentorJSON {
    id: number;
    role: MentorRoles;
    user: IUserJSON;
}

export class Mentor implements IPrototype<IMentorJSON> {
    public static fromJSON(json: IMentorJSON): Mentor {
        return new Mentor(json.id, json.role, User.fromJSON(json.user));
    }

    public readonly isLead = this.role === MentorRoles.LEAD;

    private constructor(
        public readonly id: number,
        public readonly role: MentorRoles,
        public readonly user: User
    ) {}

    public clone(overrides: Partial<IMentorJSON> = {}): Mentor {
        return Mentor.fromJSON({
            ...this,
            user: this.user.clone(),
            ...overrides
        })
    }
}
