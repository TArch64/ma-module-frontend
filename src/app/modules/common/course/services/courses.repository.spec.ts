import {Observable, of} from "rxjs";
import {CoursesRepository} from './courses.repository';
import {ICommonSeasonsService, Season} from "@common/season";
import {ICoursesRepositorySync} from "../sync";
import {CourseTypes} from "../enums";
import {Course, ICourseJSON} from "../entities";

class MockCoursesRepositorySync implements ICoursesRepositorySync {
    constructor(private courses: ICourseJSON[]) {}

    loadCourses(season: Season): Observable<ICourseJSON[]> {
        return of(this.courses);
    }
}

class MockCommonSeasonsService implements ICommonSeasonsService {
    seasons$: Observable<Season[]> = of([]);
    seasonsSnapshot: Season[] = [];
    isSeasonsLoaded: boolean = true;
    activeSeason: Season | null = null;

    currentSeason$: Observable<Season | null> = of(null);
    currentSeasonSnapshot: Season | null = null;

    loadSeasons(): Observable<Season[]> {
        return of([]);
    }

    changeCurrentSeason(season: Season): void {}
    addSeason(season: Season): void {}
    updateSeason(season: Season): void {}
    removeSeason(season: Season): void {}
}

function createSync(courses: ICourseJSON[]): ICoursesRepositorySync {
    return new MockCoursesRepositorySync(courses);
}

function createRepository(sync: ICoursesRepositorySync): CoursesRepository {
    return new CoursesRepository(sync, new MockCommonSeasonsService());
}

const createCourseJson = (attrs: Partial<ICourseJSON> = {}): ICourseJSON => ({
    id: 'test',
    name: 'test',
    type: CourseTypes.GENERAL,
    ...attrs
});

function createCourse(attrs: Partial<ICourseJSON> = {}): Course {
    return Course.fromJSON(createCourseJson(attrs));
}

describe('load courses', () => {
    test('should load courses', async () => {
        const sync = createSync([
            createCourseJson({ id: 'test-1' }),
            createCourseJson({ id: 'test-2' }),
            createCourseJson({ id: 'test-3' })
        ]);
        const repository = createRepository(sync);
        const courses = await repository.loadCourses().toPromise();

        expect(courses).toMatchSnapshot();
    });

    test('should save snapshot', async () => {
        const sync = createSync([
            createCourseJson({ id: 'test-1' }),
            createCourseJson({ id: 'test-2' }),
            createCourseJson({ id: 'test-3' })
        ]);
        const repository = createRepository(sync);
        await repository.loadCourses().toPromise();

        expect(repository.coursesSnapshot).toMatchSnapshot();
    });
});

describe('add course', () => {
    test('should add course to empty array', () => {
        const repository = createRepository(createSync([]));

        repository.addCourse(createCourse());

        expect(repository.coursesSnapshot).toMatchSnapshot();
    });

    test('should add course to existing courses', async () => {
        const repository = createRepository(createSync([
            createCourseJson({ id: 'test-1' })
        ]));
        await repository.loadCourses().toPromise();

        repository.addCourse(createCourse());

        expect(repository.coursesSnapshot).toMatchSnapshot();
    });
});
