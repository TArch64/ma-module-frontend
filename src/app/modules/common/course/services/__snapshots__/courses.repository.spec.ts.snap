// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add course should add course to empty array 1`] = `
Array [
  Course {
    "id": "test",
    "name": "test",
    "type": "GENERAL",
  },
]
`;

exports[`add course should add course to existing courses 1`] = `
Array [
  Course {
    "id": "test",
    "name": "test",
    "type": "GENERAL",
  },
  Course {
    "id": "test-1",
    "name": "test",
    "type": "GENERAL",
  },
]
`;

exports[`load courses should load courses 1`] = `
Array [
  Course {
    "id": "test-1",
    "name": "test",
    "type": "GENERAL",
  },
  Course {
    "id": "test-2",
    "name": "test",
    "type": "GENERAL",
  },
  Course {
    "id": "test-3",
    "name": "test",
    "type": "GENERAL",
  },
]
`;

exports[`load courses should save snapshot 1`] = `
Array [
  Course {
    "id": "test-1",
    "name": "test",
    "type": "GENERAL",
  },
  Course {
    "id": "test-2",
    "name": "test",
    "type": "GENERAL",
  },
  Course {
    "id": "test-3",
    "name": "test",
    "type": "GENERAL",
  },
]
`;
